From 4e2c59966c3462941d3ec9936503d59e94b4aa94 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:39:20 -0400
Subject: [alsa] pcm: Introduce protocol version field to sw_params

Message-id: <140972823573-395-patchset-jkysela@redhat.com>
Patchwork-id: 91091
O-Subject: [RHEL 7.1 PATCH 394/442] ALSA: pcm: Introduce protocol version field to sw_params
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

For controlling the new fields more strictly, add sw_params.proto
field indicating the protocol version of the user-space.  User-space
should fill the SNDRV_PCM_VERSION value it's built with, then kernel
can know whether the new fields should be evaluated or not.

And now tstamp_type field is evaluated only when the valid value is
set there.  This avoids the wrong override of tstamp_type to zero,
which is SNDRV_PCM_TSTAMP_TYPE_GETTIMEOFDAY.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: 589008106b9b64153554c8f3f120ee512b134bb5)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/uapi/sound/asound.h b/include/uapi/sound/asound.h
index 3204471..762c74b 100644
--- a/include/uapi/sound/asound.h
+++ b/include/uapi/sound/asound.h
@@ -388,8 +388,8 @@ struct snd_pcm_sw_params {
  snd_pcm_uframes_t silence_threshold; /* min distance from noise for silence filling */
  snd_pcm_uframes_t silence_size;  /* silence block size */
  snd_pcm_uframes_t boundary;  /* pointers wrap point */
- unsigned int tstamp_type;  /* timestamp type */
- int pads;    /* alignment, reserved */
+ unsigned int proto;   /* protocol version */
+ unsigned int tstamp_type;  /* timestamp type (req. proto >= 2.0.12) */
  unsigned char reserved[56];  /* reserved for future */
 };
 
diff --git a/sound/core/pcm_native.c b/sound/core/pcm_native.c
index 27de230..63f1e97 100644
--- a/sound/core/pcm_native.c
+++ b/sound/core/pcm_native.c
@@ -543,7 +543,8 @@ static int snd_pcm_sw_params(struct snd_pcm_substream *substream,
 
  if (params->tstamp_mode > SNDRV_PCM_TSTAMP_LAST)
   return -EINVAL;
- if (params->tstamp_type > SNDRV_PCM_TSTAMP_TYPE_LAST)
+ if (params->proto >= SNDRV_PROTOCOL_VERSION(2, 0, 12) &&
+     params->tstamp_type > SNDRV_PCM_TSTAMP_TYPE_LAST)
   return -EINVAL;
  if (params->avail_min == 0)
   return -EINVAL;
@@ -559,7 +560,8 @@ static int snd_pcm_sw_params(struct snd_pcm_substream *substream,
  err = 0;
  snd_pcm_stream_lock_irq(substream);
  runtime->tstamp_mode = params->tstamp_mode;
- runtime->tstamp_type = params->tstamp_type;
+ if (params->proto >= SNDRV_PROTOCOL_VERSION(2, 0, 12))
+  runtime->tstamp_type = params->tstamp_type;
  runtime->period_step = params->period_step;
  runtime->control->avail_min = params->avail_min;
  runtime->start_threshold = params->start_threshold;
-- 
1.7.1